### ISSUE TITLE
References to parameters and responses are not preserved

### ISSUE THREAD

**Issue by leochr** [swagger-api/swagger-parser#437]
References to responses and parameters are replaced with the referenced object in the serialized output. It should be similar to definitions, which preserves references.

input.yaml

<div class="highlight highlight-source-yaml"><pre><span>swagger: '2.0'
info:
  title: test
  version: "1.0.0"
  description: test
host: api.test.com
paths:
  '/buckets/{bucketKey}':
    get:
      description: test
      parameters:
      - $ref: '#/parameters/bucketKey'
      responses:
        '200':
          description: success
          schema:
            type: array
            items:
              $ref: './buckets.yaml'
    post:
      description: test
      parameters:
      - $ref: '#/parameters/bucketKey'
      responses:
        '200':
          $ref: '#/responses/objectKey'
  '/buckets/{bucketKey}/object':
    get:
      description: test
      parameters:
      - $ref: '#/parameters/bucketKey'
      responses:
        '200':
          $ref: '#/responses/objectKey'
parameters:
  bucketKey:
    name: bucketKey
    in: path
    description: Referenced parameter
    required: true
    type: string
responses:
  objectKey:
    description: Referenced response
    schema:
      type: object
      properties:
        paging:
          $ref: '#/definitions/StatusResponse'
definitions:
  StatusResponse:
    description: Referenced definition
    type: object
    properties:
      http_code:
        type: integer</span></pre></div>

buckets.yaml

<div class="highlight highlight-source-yaml"><pre><span>---
description: "Definition from external doc"
type: "object"
properties:
  hello:
    type: "string"
  world:
    type: "integer"
</span></pre></div>

output.yaml 
<div class="highlight highlight-source-yaml"><pre><span>swagger: "2.0"
info:
  description: "test"
  version: "1.0.0"
  title: "test"
host: "api.test.com"
paths:
  /buckets/{bucketKey}:
    get:
      description: "test"
      parameters:
      - name: "bucketKey"
        in: "path"
        description: "Referenced parameter"
        required: true
        type: "string"
      responses:
        200:
          description: "success"
          schema:
            type: "array"
            items:
              $ref: "#/definitions/buckets"
    post:
      description: "test"
      parameters:
      - name: "bucketKey"
        in: "path"
        description: "Referenced parameter"
        required: true
        type: "string"
      responses:
        200:
          description: "Referenced response"
          schema:
            type: "object"
            properties:
              paging:
                $ref: "#/definitions/StatusResponse"
  /buckets/{bucketKey}/object:
    get:
      description: "test"
      parameters:
      - name: "bucketKey"
        in: "path"
        description: "Referenced parameter"
        required: true
        type: "string"
      responses:
        200:
          description: "Referenced response"
          schema:
            type: "object"
            properties:
              paging:
                $ref: "#/definitions/StatusResponse"
definitions:
  StatusResponse:
    type: "object"
    properties:
      http_code:
        type: "integer"
    description: "Referenced definition"
  buckets:
    type: "object"
    properties:
      hello:
        type: "string"
      world:
        type: "integer"
    description: "Definition from external doc"
parameters:
  bucketKey:
    name: "bucketKey"
    in: "path"
    description: "Referenced parameter"
    required: true
    type: "string"
responses:
  objectKey:
    description: "Referenced response"
    schema:
      type: "object"
      properties:
        paging:
          $ref: "#/definitions/StatusResponse"</span></pre></div>

expected_output.yaml

<div class="highlight highlight-source-yaml"><pre><span>swagger: "2.0"
info:
  description: "test"
  version: "1.0.0"
  title: "test"
host: "api.test.com"
paths:
  /buckets/{bucketKey}:
    get:
      description: "test"
      parameters:
      - $ref: '#/parameters/bucketKey'
      responses:
        200:
          description: "success"
          schema:
            type: "array"
            items:
              $ref: "#/definitions/buckets"
    post:
      description: "test"
      parameters:
      - $ref: '#/parameters/bucketKey'
      responses:
        200:
          $ref: '#/responses/objectKey'
  /buckets/{bucketKey}/object:
    get:
      description: "test"
      parameters:
      - $ref: '#/parameters/bucketKey'
      responses:
        200:
          $ref: '#/responses/objectKey'
definitions:
  StatusResponse:
    type: "object"
    properties:
      http_code:
        type: "integer"
    description: "Referenced definition"
  buckets:
    type: "object"
    properties:
      hello:
        type: "string"
      world:
        type: "integer"
    description: "Definition from external doc"
parameters:
  bucketKey:
    name: "bucketKey"
    in: "path"
    description: "Referenced parameter"
    required: true
    type: "string"
responses:
  objectKey:
    description: "Referenced response"
    schema:
      type: "object"
      properties:
        paging:
          $ref: "#/definitions/StatusResponse"</span></pre></div>

As you can see, the parameter-reference **bucketEye** and response-reference **objectKey** are referenced by multiple operations. In the output, those references are replaced by the referenced object's content. It results in duplicated content within the document. Also those references under parameters and responses sections are no longer used after serialization.

For definitions, the references are kept. If the reference is external then it's resolved and copied into the document and the new internal reference is used from that point on. Parameters and responses should at least keep the internal references intact. 

A user would want to 'resolve' the external references, but still want to keep the internal references preserved. Ideally preserving the internal references should be the default behaviour, so the document doesn't get bloated by copying the referenced object's content to each instance.

**Comment by leochr** [comment]
@fehguy, opening this issue as per our discussion earlier.

**Comment by leochr** [comment]
@fehguy you are probably occupied with other components. but wondering if you had a chance to think about this one. thanks.

**Comment by fehguy** [comment]
Yes, for the new parser, I introduced a [ParseOptions.java](https://github.com/swagger-api/swagger-parser/blob/feature/3.0.0-rc0/modules/swagger-parser-core/src/main/java/io/swagger/parser/models/ParseOptions.java) class to hold options for how we resolve or don't.  I suggest that we enumerate the parse options in the class so users have full control over how the specification is resolved.

**Comment by chandan485** [comment]
I did not find the above mentioned Java class. On which version did you add this class? Has this issue resolved??

**Comment by gracekarina** [comment]
@chandan485  https://github.com/swagger-api/swagger-parser/blob/2.0/modules/swagger-parser-core/src/main/java/io/swagger/v3/parser/core/models/ParseOptions.java

**Comment by chandan485** [comment]
Thanks for quick response.
Does this support for swagger 2.0? I am using swagger-parser V 1.0.33 . From which class of swagger-parser library do i need to use this ParseOption.java?

**Comment by gracekarina** [comment]
This is for openAPI 3.0

**Comment by anthomps** [comment]
To clarify: can the improved tools for OpenAPI 3.0 also process specifications in 2.0 format or are they mutually exclusive? i.e. I have to use 2.0 tools with a 2.0 spec?

Are any fixes being back ported to the 2.0 tools?

**Comment by gracekarina** [comment]
Hi, @anthomps @chandan485  sorry for the delay reply OpenApiParser can be use to parse swagger 1.x, 2.0 and 3.0., parser will internally convert the previous version to the 3.0 version of the spec.
